### PR TITLE
DropDownButton: add some SSR tests

### PR DESCRIPTION
**Pull request by ksercs**

### CODE REVIEW

File: testing/tests/DevExpress.serverSide/dropDownButton.tests.js [hunk 1]
```diff
@@ -0,0 +1 @@
+require('../DevExpress.ui.widgets/dropDownButton.base.tests.js');
```
Comment (DokaRus) [inline testing/tests/DevExpress.serverSide/dropDownButton.tests.js]:
```suggestion
import '../DevExpress.ui.widgets/dropDownButton.base.tests.js';
```

**Review by DokaRus — COMMENTED** [review]
I think dropDownButton.tests should be extend by the "optionChange" tests module. Some of tests can have a different count of assertions in SSR and client mode (you can check the `core/utils/window => hasWindow` state)

File: testing/tests/DevExpress.serverSide/dropDownButton.tests.js [hunk 2]
```diff
@@ -0,0 +1 @@
+import '../DevExpress.ui.widgets/dropDownButton.base.tests.js';
```
Comment (nightskylark) [inline testing/tests/DevExpress.serverSide/dropDownButton.tests.js]: We named same tests for other widgets as `*.markup.tests.js`. Why did you use another name?

Comment (ksercs) [reply testing/tests/DevExpress.serverSide/dropDownButton.tests.js]: Because there are not only tests for markup in this file, but also option change tests, etc. But I agree that I should still name it in accordance with tests for other widgets

**Review by San4es — COMMENTED** [review]
Some tests are failed with:
"Error: Window fields using is prevented"

File: js/ui/drop_down_button.js [hunk 3]
```diff
@@ -331,7 +331,7 @@ const DropDownButton = Widget.inherit({
                     y: -1
                 }
             }
-        }, this._options.cache('dropDownOptions'));
+        }, this._options.cache('dropDownOptions'), { visible: this.option('opened') });
```
Comment (San4es) [inline js/ui/drop_down_button.js:337]: Is this change necessary? What test covers this?

File: js/ui/drop_down_button.js [hunk 4]
```diff
@@ -519,14 +523,18 @@ const DropDownButton = Widget.inherit({
...
                 break;
             case 'keyExpr':
                 this._compileKeyGetter();
+                this._setListOption(name, value);
                 break;
             case 'buttonGroupOptions':
                 this._innerWidgetOptionChanged(this._buttonGroup, args);
                 break;
             case 'dropDownOptions':
+                args.value.visible && delete args.value.visible;
```
Comment (San4es) [inline js/ui/drop_down_button.js]: Is this change necessary? What test covers this?

Comment (ksercs) [reply js/ui/drop_down_button.js]: https://github.com/DevExpress/DevExtreme/pull/11847/files#diff-d9fa2b0f966e8a8b29e5b1d170e9ea05R599

File: js/ui/drop_down_button.js (hunk 3 above)
Comment (ksercs) [reply js/ui/drop_down_button.js:337]: https://github.com/DevExpress/DevExtreme/pull/11847/files#diff-0360e44628ce8bae07db8d6360fc7ba4R103

File: js/ui/drop_down_button.js [hunk 5]
```diff
@@ -424,6 +424,10 @@ const DropDownButton = Widget.inherit({
     },
 
     toggle(visible) {
+        if(!windowUtils.hasWindow()) {
```
Comment (San4es) [inline js/ui/drop_down_button.js]: Let's get rid of this and use hasWidow() check here:
https://github.com/DevExpress/DevExtreme/pull/11847/files#diff-cc46f04faa146e60076fc47eff5a47c1R309

File: js/ui/drop_down_button.js [hunk 6]
```diff
@@ -588,7 +598,7 @@ const DropDownButton = Widget.inherit({
                 this._createSelectionChangedAction();
                 break;
             case 'deferRendering':
-                if(!value && !this._popup) {
+                if(windowUtils.hasWindow() && !value && !this._popup) {
```
Comment (San4es) [inline js/ui/drop_down_button.js]: See this comment:
https://github.com/DevExpress/DevExtreme/pull/11847/files#diff-cc46f04faa146e60076fc47eff5a47c1R427

File: js/ui/drop_down_button.js (hunk 3 above)
Comment (San4es) [reply js/ui/drop_down_button.js:337]: Do we need to do this for `deferRendering` as well?

File: js/ui/drop_down_button.js [hunk 7]
```diff
@@ -588,10 +597,8 @@ const DropDownButton = Widget.inherit({
                 this._createSelectionChangedAction();
                 break;
             case 'deferRendering':
-                if(!value && !this._popup) {
-                    this._renderPopup();
-                    this._renderContent();
-                }
+                this._renderPopup();
```
Comment (San4es) [inline js/ui/drop_down_button.js]: Do we need to render popup if it is already rendered? To answer this question, let's cover this case by a separate test.

File: js/ui/drop_down_button.js (hunk 3 above)
Comment (ksercs) [reply js/ui/drop_down_button.js:337]: No, `deferRendering` works correct, because `dropDownOptions.deferRendering` doesn't affect on popup:
- `deferRendering` is true: because the popup widget has not been created yet. Option will affect only on popup init.
- `deferRendering` is false: because the popup widget has already been created

File: js/ui/drop_down_button.js (hunk 5 above)
Comment (ksercs) [reply js/ui/drop_down_button.js]: fixed

File: js/ui/drop_down_button.js (hunk 6 above)
Comment (ksercs) [reply js/ui/drop_down_button.js]: fixed

File: js/ui/drop_down_button.js (hunk 7 above)
Comment (ksercs) [reply js/ui/drop_down_button.js]: fixed it

File: testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js [hunk 8]
```diff
@@ -633,6 +633,16 @@ QUnit.module('common use cases', {
         assert.strictEqual(getActionButton(this.dropDownButton).text(), 'Trial for Visual Studio', 'action button has been changed');
     });
 
+    QUnit.test('deferRendering should not do anything if popup has already been rendered', function(assert) {
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js:636]: Wouldn't it be better to check the popup element strict equality between deferRendering changes to avoid private method usage?

**Review by San4es — DISMISSED** [review]
LGTM

File: testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js (hunk 8 above)
Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js:636]: It's good idea, thanks

File: js/ui/drop_down_button.js [hunk 9]
```diff
@@ -519,14 +522,20 @@ const DropDownButton = Widget.inherit({
...
             case 'keyExpr':
                 this._compileKeyGetter();
+                this._setListOption(name, value);
                 break;
             case 'buttonGroupOptions':
                 this._innerWidgetOptionChanged(this._buttonGroup, args);
                 break;
             case 'dropDownOptions':
+                if(args.value.visible !== undefined) {
+                    delete args.value.visible;
```
Comment (DokaRus) [inline js/ui/drop_down_button.js:540]: `instance.option('dropDownOptions.visible', true)` raises `valueChanged` event with args: 
```
{
  name: "dropDownOptions",
  fullName: "dropDownOptions.visible",
  value: true
}

Comment (ksercs) [reply js/ui/drop_down_button.js:540]: fixed it, thanks

File: testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js (hunk 8 above)
Comment (San4es) [reply testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js:636]: I meant popup elements :)

Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js:636]: I understood, and even tested it in the playground, but it didn't stop me from screwing up :-D

File: testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js [hunk 10]
```diff
@@ -794,6 +633,16 @@ QUnit.module('common use cases', {
...
+        const $dropDownButton = this.dropDownButton._popup.$element();
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js]: What about using `getPopup`:
https://github.com/DevExpress/DevExtreme/blob/ebc39a2da672b7e3dfd16eabe72012f1548896b1/testing/tests/DevExpress.ui.widgets/dropDownButton.tests.js#L28-L30

**Review by San4es — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]
